src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1312,12 +1322,17 @@ internal static bool NeedsAnalysis(CSharpCompilation compilation)
             if (compilation.LanguageVersion < MessageID.IDS_FeatureNullableReferenceTypes.RequiredVersion() || !compilation.ShouldRunNullableWalker)
             {
 #if DEBUG
-                // Always run analysis in debug builds so that we can more reliably catch
-                // nullable regressions e.g. https://github.com/dotnet/roslyn/issues/40136
-                diagnostics = new DiagnosticBag();
-#else
-                return;
+                if (ShouldRunAnalysisInDebug(compilation))
+                {
+                    // Always run analysis in debug builds so that we can more reliably catch
+                    // nullable regressions e.g. https://github.com/dotnet/roslyn/issues/40136
```

Comment:
This comment seems more applicable on the method now than the invocation of the method.